cmd/zt_remove_blob_test.go

```diff
@@ -714,3 +718,101 @@ func (s *cmdIntegrationSuite) TestRemoveBlobsUnderVirtualDirWithFromTo(c *chk.C)
+
+	// From FE, 30 seconds is guaranteed to be enough.
+	time.Sleep(time.Second * 30)
+
+	// set up the container with numerous blobs
+	containerURL, containerName := createNewContainer(c, serviceURL)
+	defer deleteContainer(c, containerURL)
+	blobName, blobList, _ := scenarioHelper{}.generateCommonRemoteScenarioForSoftDelete(c, containerURL, "")
+	c.Assert(containerURL, chk.NotNil)
+	c.Assert(len(blobList), chk.Equals, 3)
+
+	list, _ := containerURL.ListBlobsFlatSegment(ctx, azblob.Marker{}, azblob.ListBlobsSegmentOptions{Details: azblob.BlobListingDetails{Deleted: true, Snapshots: true}, Prefix: blobName})
+	c.Assert(list.Segment.BlobItems, chk.NotNil)
+	c.Assert(len(list.Segment.BlobItems), chk.Equals, 4)
+
+	// set up interceptor
+	mockedRPC := interceptor{}
+	Rpc = mockedRPC.intercept
+	mockedRPC.init()
+
+	// construct the raw input to simulate user input
+	rawBlobURLWithSAS := scenarioHelper{}.getRawBlobURLWithSAS(c, containerName, blobName)
+	raw := getDefaultRemoveRawInput(rawBlobURLWithSAS.String())
+	raw.recursive = true
+	raw.permanentDeleteOption = "snapshotsandversions"
+	runCopyAndVerify(c, raw, func(err error) {
+		c.Assert(err, chk.IsNil)
+
+		// validate that the right number of transfers were scheduled
+		c.Assert(len(mockedRPC.transfers), chk.Equals, 3)
+	})
+}
+
+func (s *cmdIntegrationSuite) TestPermDeleteSnapshotsVersionsUnderContainer(c *chk.C) {
+	serviceURL := setUpAccountPermDelete(c)
+
+	// From FE, 30 seconds is guaranteed to be enough.
+	time.Sleep(time.Second * 30)
+
+	// set up the container with numerous blobs
+	containerURL, containerName := createNewContainer(c, serviceURL)
+	defer deleteContainer(c, containerURL)
+	_, blobList, listOfTransfers := scenarioHelper{}.generateCommonRemoteScenarioForSoftDelete(c, containerURL, "")
+	c.Assert(containerURL, chk.NotNil)
+	c.Assert(len(blobList), chk.Equals, 3)
+
+	// set up interceptor
+	mockedRPC := interceptor{}
+	Rpc = mockedRPC.intercept
+	mockedRPC.init()
+
+	// construct the raw input to simulate user input
+	rawContainerURLWithSAS := scenarioHelper{}.getRawContainerURLWithSAS(c, containerName)
+	raw := getDefaultRemoveRawInput(rawContainerURLWithSAS.String())
+	raw.recursive = true
+	raw.permanentDeleteOption = "snapshotsandversions"
+	runCopyAndVerify(c, raw, func(err error) {
+		c.Assert(err, chk.IsNil)
+
+		// validate that the right number of transfers were scheduled
+		c.Assert(len(mockedRPC.transfers), chk.Equals, len(listOfTransfers))
+	})
+}
+
+func setUpAccountPermDelete(c *chk.C) azblob.ServiceURL {
+	accountName, accountKey := getAccountAndKey()
+	u, _ := url.Parse(fmt.Sprintf("https://%s.blob.core.windows.net/", accountName))
+
+	credential, err := azblob.NewSharedKeyCredential(accountName, accountKey)
+	if err != nil {
+		panic(err)
+	}
+
+	sasQueryParams, err := azblob.AccountSASSignatureValues{
+		Protocol:      azblob.SASProtocolHTTPS,
+		ExpiryTime:    time.Now().UTC().Add(168 * time.Hour), // week long sas
```

Comment:
We actually have a 30m or so maximum length on testing; 6-12 hours may actually be a much better fit here in case some bizarre leakage of the SAS token happens in testing.